be/src/storage/storage_engine.cpp

```diff
@@ -543,6 +544,43 @@ void StorageEngine::_start_clean_fd_cache() {
     VLOG(10) << "Cleaned file descriptor cache";
 }
 
+void StorageEngine::compaction_check() {
+    int checker_one_round_sleep_time_s = 1800;
+    while (true) {
+        MonotonicStopWatch stop_watch;
+        stop_watch.start();
+        LOG(INFO) << "start to check compaction";
+        size_t num = compaction_check_one_round();
+        stop_watch.stop();
+        LOG(INFO) << num << " tablets checked. time elapse:" << stop_watch.elapsed_time() / 1e9 << " seconds."
+                  << " compaction checker will be scheduled again in " << checker_one_round_sleep_time_s << " seconds";
+        std::this_thread::sleep_for(std::chrono::seconds(checker_one_round_sleep_time_s));
+    }
+}
+
+// Base compaction may be started by time(once every day now)
+// Compaction checker will check whether to schedule base compaction for tablets
+size_t StorageEngine::compaction_check_one_round() {
+    size_t batch_size = 1000;
+    int batch_sleep_time_ms = 1000;
+    std::vector<TabletSharedPtr> tablets;
+    tablets.reserve(batch_size);
+    size_t tablets_num_checked = 0;
+    while (true) {
+        bool finished = _tablet_manager->get_next_batch_tablets(batch_size, &tablets);
+        for (auto& tablet : tablets) {
+            _compaction_manager->update_tablet(tablet, true, false);
+        }
+        tablets_num_checked += tablets.size();
+        tablets.clear();
+        if (finished) {
+            break;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(batch_sleep_time_ms));
```

Comment:
one second sleep seems not enough

Reply:
it is used to control the speed of checker, and do not use cpu that much. It does not limit the candidate queue size.